File: 623-add-one-row-to-tree/623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        
        if depth == 1:
            newRoot = TreeNode(val, root)
            return newRoot
        
        currDepth = 0
        queue = deque()
        queue.append(root)
        
        isDone = False
        while queue or not isDone:
            levelSize = len(queue)
            currDepth += 1
            
            for _ in range(levelSize):
                currNode = queue.popleft()
                
                if currDepth == depth - 1:
                    leftNode = TreeNode(val)
                    rightNode = TreeNode(val)
                    
                    if currNode.left:
                        leftNode.left = currNode.left
                    if currNode.right:
                        rightNode.right = currNode.right
                    
                    currNode.left = leftNode
                    currNode.right = rightNode
                    
                    isDone = True
                
                if currNode.left:
                    queue.append(currNode.left)
                
                if currNode.right:
                    queue.append(currNode.right)
        return root
                    
        

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        return solve(root, val, depth);
    }
private:
    TreeNode* solve(TreeNode* root, int val, int depth) {
        queue<TreeNode*> queue;
        queue.push(root);
        
        
        // if depth is 1 then new root will be changed to given val
        if(depth == 1) {
            TreeNode* newRoot = new TreeNode(val, root, nullptr);
            return newRoot;
        }
        
        int currDepth = 0;
        
        // is addition of row done
        bool isDone = false;
        
        while(!queue.empty() || !isDone) {
            int levelSize = queue.size();
            ++currDepth;
            while(levelSize--) {
                auto currNode = queue.front();
                queue.pop();
                
                // if the next level is the required depth
                if(currDepth == depth - 1) {
                    // creating tree nodes
                    TreeNode* leftNode = new TreeNode(val);
                    TreeNode* rightNode = new TreeNode(val);
                    
                    // preserving the left and right subtree
                    if(currNode -> left) leftNode -> left = currNode -> left;
                    if(currNode -> right) rightNode -> right = currNode -> right;
                    
                    // assigning new row to the left and right pointers
                    currNode -> left = leftNode;
                    currNode -> right = rightNode;
                    isDone = true;
                }
                if(currNode -> left) queue.push(currNode -> left);
                if(currNode -> right) queue.push(currNode -> right);
            }
        }
        return root;
    }
};